[Windows SEH] Crash on Windows SEH on 19.1.0

Stack dump:
0.      Program arguments: C:\\PROGRA~1\\LLVM\\bin\\clang-cl.exe /nologo -TP -D_USE_MATH_DEFINES -D_VARIADIC_MAX=10 -D_WIN32_WINNT=0x0601 -D__OPENCV_BUILD=1 -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -IC:\\Users\\rysza\\lib\\opencv\\build\\3rdparty\\ippicv\\ippicv_win\\icv\\include -IC:\\Users\\rysza\\lib\\opencv\\build\\3rdparty\\ippicv\\ippicv_win\\iw\\include -IC:\\Users\\rysza\\lib\\opencv\\build -IC:\\Users\\rysza\\lib\\opencv\\modules\\ts\\include -IC:\\Users\\rysza\\lib\\opencv\\build\\modules\\ts -IC:\\Users\\rysza\\lib\\opencv\\modules\\core\\include -IC:\\Users\\rysza\\lib\\opencv\\modules\\imgproc\\include -IC:\\Users\\rysza\\lib\\opencv\\modules\\imgcodecs\\include -IC:\\Users\\rysza\\lib\\opencv\\modules\\videoio\\include -IC:\\Users\\rysza\\lib\\opencv\\modules\\highgui\\include "-imsvcC:\\Program Files (x86)\\Intel\\oneAPI\\mkl\\latest\\include" /DWIN32 /D_WINDOWS /W4 /GR /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi /fp:precise -W -Wreturn-type -Wnon-virtual-dtor -Waddress -Wsequence-point -Wformat -Wformat-security -Wmissing-declarations -Wmissing-prototypes -Wstrict-prototypes -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Winconsistent-missing-override -Wno-delete-non-virtual-dtor -Wno-unnamed-type-template-args -Wno-comment -Wno-deprecated-enum-enum-conversion -Wno-deprecated-anon-enum-enum-conversion -Wno-long-long -Qunused-arguments /FS -msse3 -mssse3 -msse4.1 -mpopcnt -msse4.2 -mavx -mf16c -mavx2 -mfma /EHa /wd4127 /wd4251 /wd4324 /wd4275 /wd4512 /wd4589 /wd4819 -Wno-undef /O2 /Ob2 /DNDEBUG -DNDEBUG -MD /showIncludes /Fomodules\\ts\\CMakeFiles\\opencv_ts.dir\\src\\ts_gtest.cpp.obj /Fdlib\\opencv_ts4100.pdb -c -- C:\\Users\\rysza\\lib\\opencv\\modules\\ts\\src\\ts_gtest.cpp
1.      <eof> parser at end of file
2.      Code generation
3.      Running pass 'Function Pass Manager' on module 'C:\Users\rysza\lib\opencv\modules\ts\src\ts_gtest.cpp'.
4.      Running pass 'X86 Assembly Printer' on function '@"??$HandleSehExceptionsInMethodIfSupported@VTest@testing@@X@internal@testing@@YAXPEAVTest@1@P821@EAAXXZPEBD@Z"'
Exception Code: 0xC0000005
 #0 0x00007ff6ad89522c (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x1cc522c)
 #1 0x00007ff6ad894df1 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x1cc4df1)
 #2 0x00007ff6ac2f9eea (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x729eea)
 #3 0x00007ff6ad6a2293 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x1ad2293)
 #4 0x00007ff6ac13f7ed (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x56f7ed)
 #5 0x00007ff6ad257001 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x1687001)
 #6 0x00007ff6ad2416b1 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x16716b1)
 #7 0x00007ff6abea3741 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x2d3741)
 #8 0x00007ff6abea30e7 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x2d30e7)
 #9 0x00007ff6abe9c8ac (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x2cc8ac)
#10 0x00007ff6ad23bfc5 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x166bfc5)
#11 0x00007ff6ad72e188 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x1b5e188)
#12 0x00007ff6ac18e341 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x5be341)
#13 0x00007ff6ac18ded2 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x5bded2)
#14 0x00007ff6ac18ac46 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x5bac46)
#15 0x00007ff6ac187c05 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x5b7c05)
#16 0x00007ff6ac18659e (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x5b659e)
#17 0x00007ff6ac8a003d (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0xcd003d)
#18 0x00007ff6ac310075 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x740075)
#19 0x00007ff6ac30fe68 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x73fe68)
#20 0x00007ff6ac00c008 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x43c008)
#21 0x00007ff6ac00b38e (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x43b38e)
#22 0x00007ff6ac00af07 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x43af07)
#23 0x00007ff6ac002696 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x432696)
#24 0x00007ff6abfffe13 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x42fe13)
#25 0x00007ff6ad7ce458 (C:\PROGRA~1\LLVM\bin\clang-cl.exe+0x1bfe458)
#26 0x00007fff5932257d (C:\Windows\System32\KERNEL32.DLL+0x1257d)
#27 0x00007fff5a6caf28 (C:\Windows\SYSTEM32\ntdll.dll+0x5af28)
clang-cl: error: clang frontend command failed due to signal (use -v to see invocation)
clang version 19.1.0
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\LLVM\bin
[ts_gtest-cd8cab.zip](https://github.com/user-attachments/files/17089857/ts_gtest-cd8cab.zip)

The function is here (I deleted some ifdefs):
```
template <class T, typename Result>
Result HandleSehExceptionsInMethodIfSupported(
    T* object, Result (T::*method)(), const char* location) {
  __try {
    return (object->*method)();
  } __except (internal::UnitTestOptions::GTestShouldProcessSEH(  // NOLINT
      GetExceptionCode())) {
    // We create the exception message on the heap because VC++ prohibits
    // creation of objects with destructors on stack in functions using __try
    // (see error C2712).
    std::string* exception_message = FormatSehExceptionMessage(
        GetExceptionCode(), location);
    internal::ReportFailureInUnknownLocation(TestPartResult::kFatalFailure,
                                             *exception_message);
    delete exception_message;
    return static_cast<Result>(0);
  }
}
```
It is quite deep in gtest. I think this is what calls it:
```
// Creates the test object.
  Test* const test = internal::HandleExceptionsInMethodIfSupported(
      factory_, &internal::TestFactoryBase::CreateTest,
      "the test fixture's constructor");
```
I'll need try to isolate what is needed to run this.

Right I missed the delete inside the __except. That's a weird interaction. The code in __except is not protected against SEH exceptions after all. 
>     // We create the exception message on the heap because VC++ prohibits
>     // creation of objects with destructors on stack in functions using __try
>     // (see error C2712). 
C2712 is quite interesting in itself. So your example is considered ill-formed by MSVC. But the Gtest function causing that explicitly avoids the issue by not using objects that need to be destructed.

--- a/clang/lib/CodeGen/CGCleanup.cpp
+++ b/clang/lib/CodeGen/CGCleanup.cpp
@@ -159,6 +159,7 @@ void *EHScopeStack::pushCleanup(CleanupKind Kind, size_t Size) {
   bool IsEHCleanup = Kind & EHCleanup;
   bool IsLifetimeMarker = Kind & LifetimeMarker;
   bool IsFakeUse = Kind & FakeUse;
+  bool IsSEHFinallyCleanup = Kind & SEHFinallyCleanup;
 
   // Per C++ [except.terminate], it is implementation-defined whether none,
   // some, or all cleanups are called before std::terminate. Thus, when
@@ -183,6 +184,8 @@ void *EHScopeStack::pushCleanup(CleanupKind Kind, size_t Size) {
     Scope->setLifetimeMarker();
   if (IsFakeUse)
     Scope->setFakeUse();
+  if (IsSEHFinallyCleanup)
+    Scope->setSEHFinallyCleanup();
 
   // With Windows -EHa, Invoke llvm.seh.scope.begin() for EHCleanup
   // If exceptions are disabled/ignored and SEH is not in use, then there is no
@@ -191,7 +194,8 @@ void *EHScopeStack::pushCleanup(CleanupKind Kind, size_t Size) {
   // consistent with MSVC's behavior, except in the presence of -EHa.
   // Check getInvokeDest() to generate llvm.seh.scope.begin() as needed.
   if (CGF->getLangOpts().EHAsynch && IsEHCleanup && !IsLifetimeMarker &&
-      CGF->getTarget().getCXXABI().isMicrosoft() && CGF->getInvokeDest())
+      !IsSEHFinallyCleanup && CGF->getTarget().getCXXABI().isMicrosoft() &&
+      CGF->getInvokeDest())
     CGF->EmitSehCppScopeBegin();
 
   return Scope->getCleanupBuffer();
@@ -784,7 +788,7 @@ void CodeGenFunction::PopCleanupBlock(bool FallthroughIsBranchThrough,
 
   // Under -EHa, invoke seh.scope.end() to mark scope end before dtor
   bool IsEHa = getLangOpts().EHAsynch && !Scope.isLifetimeMarker();
-  const EHPersonality &Personality = EHPersonality::get(*this);
+  bool IsSEHFinallyCleanup = Scope.isSEHFinallyCleanup();
   if (!RequiresNormalCleanup) {
     // Mark CPP scope end for passed-by-value Arg temp
     //   per Windows ABI which is "normally" Cleanup in callee
@@ -794,7 +798,7 @@ void CodeGenFunction::PopCleanupBlock(bool FallthroughIsBranchThrough,
       // block.
       if (NormalDeactivateOrigIP.isSet())
         Builder.restoreIP(NormalDeactivateOrigIP);
-      if (Personality.isMSVCXXPersonality() && Builder.GetInsertBlock())
+      if (Builder.GetInsertBlock() && !IsSEHFinallyCleanup)
         EmitSehCppScopeEnd();
       if (NormalDeactivateOrigIP.isSet())
         NormalDeactivateOrigIP = Builder.saveAndClearIP();
@@ -810,10 +814,10 @@ void CodeGenFunction::PopCleanupBlock(bool FallthroughIsBranchThrough,
 
       // mark SEH scope end for fall-through flow
       if (IsEHa && getInvokeDest()) {
-        if (Personality.isMSVCXXPersonality())
-          EmitSehCppScopeEnd();
-        else
+        if (Scope.isSEHFinallyCleanup())
           EmitSehTryScopeEnd();
+        else
+          EmitSehCppScopeEnd();
       }
 
       destroyOptimisticNormalEntry(*this, Scope);
@@ -852,10 +856,10 @@ void CodeGenFunction::PopCleanupBlock(bool FallthroughIsBranchThrough,
 
       // intercept normal cleanup to mark SEH scope end
       if (IsEHa && getInvokeDest()) {
-        if (Personality.isMSVCXXPersonality())
-          EmitSehCppScopeEnd();
-        else
+        if (Scope.isSEHFinallyCleanup())
           EmitSehTryScopeEnd();
+        else
+          EmitSehCppScopeEnd();
       }
 
       // III.  Figure out where we're going and build the cleanup
@@ -1055,7 +1059,8 @@ void CodeGenFunction::PopCleanupBlock(bool FallthroughIsBranchThrough,
       EHStack.pushTerminate();
       PushedTerminate = true;
     } else if (IsEHa && getInvokeDest()) {
-      EmitSehCppScopeEnd();
+      if (!IsSEHFinallyCleanup)
+        EmitSehCppScopeEnd();
     }
 
     // We only actually emit the cleanup code if the cleanup is either

--- a/clang/lib/CodeGen/CGCleanup.h
+++ b/clang/lib/CodeGen/CGCleanup.h
@@ -99,6 +99,9 @@ protected:
     LLVM_PREFERRED_TYPE(bool)
     unsigned TestFlagInEHCleanup : 1;
 
+    LLVM_PREFERRED_TYPE(bool)
+    unsigned IsSEHFinallyCleanup : 1;
+
     /// The amount of extra storage needed by the Cleanup.
     /// Always a multiple of the scope-stack alignment.
     unsigned CleanupSize : 12;
@@ -357,6 +360,7 @@ public:
     CleanupBits.IsActive = true;
     CleanupBits.IsLifetimeMarker = false;
     CleanupBits.IsFakeUse = false;
+    CleanupBits.IsSEHFinallyCleanup = false;
     CleanupBits.TestFlagInNormalCleanup = false;
     CleanupBits.TestFlagInEHCleanup = false;
     CleanupBits.CleanupSize = cleanupSize;
@@ -391,6 +395,9 @@ public:
 
   bool isFakeUse() const { return CleanupBits.IsFakeUse; }
   void setFakeUse() { CleanupBits.IsFakeUse = true; }
+
+  bool isSEHFinallyCleanup() const { return CleanupBits.IsSEHFinallyCleanup; }
+  void setSEHFinallyCleanup() { CleanupBits.IsSEHFinallyCleanup = true; }
 
   bool hasActiveFlag() const { return ActiveFlag.isValid(); }
   Address getActiveFlag() const {

--- a/clang/lib/CodeGen/CGException.cpp
+++ b/clang/lib/CodeGen/CGException.cpp
@@ -2181,7 +2181,8 @@ llvm::Value *CodeGenFunction::EmitSEHAbnormalTermination() {
 
 void CodeGenFunction::pushSEHCleanup(CleanupKind Kind,
                                      llvm::Function *FinallyFunc) {
-  EHStack.pushCleanup<PerformSEHFinally>(Kind, FinallyFunc);
+  EHStack.pushCleanup<PerformSEHFinally>(
+      static_cast<CleanupKind>(Kind | SEHFinallyCleanup), FinallyFunc);
 }
 
 void CodeGenFunction::EnterSEHTryStmt(const SEHTryStmt &S) {
@@ -2193,7 +2194,8 @@ void CodeGenFunction::EnterSEHTryStmt(const SEHTryStmt &S) {
         HelperCGF.GenerateSEHFinallyFunction(*this, *Finally);
 
     // Push a cleanup for __finally blocks.
-    EHStack.pushCleanup<PerformSEHFinally>(NormalAndEHCleanup, FinallyFunc);
+    EHStack.pushCleanup<PerformSEHFinally>(NormalAndEHSEHFinallyCleanup,
+                                           FinallyFunc);
     return;
   }
 

--- a/clang/lib/CodeGen/CGExprCXX.cpp
+++ b/clang/lib/CodeGen/CGExprCXX.cpp
@@ -1720,6 +1720,18 @@ llvm::Value *CodeGenFunction::EmitCXXNewExpr(const CXXNewExpr *E) {
   llvm::Instruction *cleanupDominator = nullptr;
   if (E->getOperatorDelete() &&
       !E->getOperatorDelete()->isReservedGlobalPlacementOperator()) {
+    // A potentially-throwing constructor inside __try requires C++ object
+    // unwinding, which is incompatible with SEH.
+    if (getLangOpts().CXXExceptions && currentFunctionUsesSEHTry()) {
+      if (const auto *ConstructExpr = E->getConstructExpr()) {
+        const auto *FPT = ConstructExpr->getConstructor()
+                              ->getType()
+                              ->castAs<FunctionProtoType>();
+        if (!FPT->isNothrow())
+          getContext().getDiagnostics().Report(E->getBeginLoc(),
+                                               diag::err_seh_object_unwinding);
+      }
+    }
     EnterNewDeleteCleanup(*this, E, TypeIdentityArg, allocation, allocSize,
                           allocAlign, allocatorArgs);
     operatorDeleteCleanup = EHStack.stable_begin();

--- a/clang/lib/CodeGen/EHScopeStack.h
+++ b/clang/lib/CodeGen/EHScopeStack.h
@@ -92,6 +92,9 @@ enum CleanupKind : unsigned {
   // markers chiefly to be ignored in most contexts.
   FakeUse = 0x10,
   NormalFakeUse = FakeUse | NormalCleanup,
+
+  SEHFinallyCleanup = 0x20,
+  NormalAndEHSEHFinallyCleanup = SEHFinallyCleanup | NormalAndEHCleanup,
 };
 
 /// A stack of scopes which respond to exceptions, including cleanups

--- a/clang/test/CodeGen/windows-seh-EHa-TryInFinally.cpp
+++ b/clang/test/CodeGen/windows-seh-EHa-TryInFinally.cpp
@@ -43,11 +43,11 @@ int main() {
 }
 
 // CHECK-LABEL:@"?foo@@YAXXZ"()
-// CHECK: invoke.cont:
 // CHECK: invoke void @llvm.seh.try.begin()
+// CHECK: invoke.cont:
 // CHECK: store volatile i32 1, ptr %cleanup.dest.slot
 // CHECK: invoke void @llvm.seh.try.end()
-// CHECK: invoke.cont2:
+// CHECK: invoke.cont1:
 // CHECK: %cleanup.dest = load i32, ptr %cleanup.dest.slot
 // CHECK: %1 = icmp ne i32 %cleanup.dest, 0
 // CHECK: %2 = zext i1 %1 to i8
@@ -69,14 +69,13 @@ void foo()
 }
 
 // CHECK-LABEL:@"?bar@@YAHXZ"()
-// CHECK: invoke.cont:
 // CHECK: invoke void @llvm.seh.try.begin()
-// CHECK: invoke.cont1:
+// CHECK: invoke.cont:
 // CHECK: store volatile i32 1, ptr %cleanup.dest.slot
 // CHECK: invoke void @llvm.seh.try.end()
-// CHECK: invoke.cont2:
+// CHECK: invoke.cont1:
 // CHECK: call void @"?fin$0@0@bar@@"
-// CHECK: %cleanup.dest3 = load i32, ptr %cleanup.dest.slot
+// CHECK: %cleanup.dest2 = load i32, ptr %cleanup.dest.slot
 // CHECK: return:
 // CHECK: ret i32 11
 int bar()
@@ -88,5 +87,117 @@ int bar()
     if (_abnormal_termination()) {
       x = 9;
     }
+  }
+}
+
+// CHECK-LABEL: @"?foo1@@YAXXZ"()
+//
+// CHECK-NOT: invoke void @llvm.seh.scope.begin()
+// CHECK:     invoke void @llvm.seh.try.begin()
+//
+// CHECK-NOT: invoke void @llvm.seh.try.begin()
+// CHECK:     invoke.cont1:
+//
+// CHECK:     invoke void @llvm.seh.scope.begin()
+// CHECK:     invoke noundef ptr @"??0A@@QEAA@XZ"
+//
+// CHECK:     invoke void @llvm.seh.scope.end()
+//
+// CHECK-NOT: @llvm.seh
+//
+// CHECK:     invoke void @llvm.seh.scope.begin()
+//
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     invoke void @"??1A@@QEAA@XZ"
+//
+// CHECK:     invoke void @llvm.seh.scope.end()
+// CHECK-NOT: @llvm.seh
+// CHECK:     br label %delete.end
+//
+// CHECK:     invoke void @llvm.seh.try.end()
+//
+// CHECK-NOT: invoke void @llvm.seh.try.end()
+// CHECK:     invoke.cont15:
+//
+// CHECK:     call void @"?fin$0@0@foo1@@"(i8 noundef 0
+// CHECK-NOT: @llvm.seh
+//
+// CHECK:     ehcleanup:
+// CHECK:     invoke void @llvm.seh.scope.end()
+// CHECK:     invoke.cont7:
+// CHECK:     cleanupret
+//
+// CHECK:     invoke void @llvm.seh.scope.end()
+// CHECK:     invoke.cont14:
+// CHECK:     cleanupret
+//
+// CHECK:     ehcleanup16:
+// CHECK:     call void @"?fin$0@0@foo1@@"(i8 noundef 1
+// CHECK-NOT: @llvm.seh
+// CHECK:     cleanupret
+struct A {
+	A() noexcept;
+	~A();
+};
+
+void foo1() {
+	__try {
+		A* a = new A;
+		delete a;
+	}
+	__finally {
+	}
+}
+
+// CHECK-LABEL: @"?seh_finally_no_obj@@YAXXZ"()
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     invoke void @llvm.seh.try.begin()
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     invoke void @"?might_throw@@YAXXZ"()
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     invoke void @llvm.seh.try.end()
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     call void @"?fin$0@0@seh_finally_no_obj@@"(i8 noundef 0
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     call void @"?fin$0@0@seh_finally_no_obj@@"(i8 noundef 1
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     cleanupret
+void might_throw();
+void seh_finally_no_obj() {
+  __try {
+    might_throw();
+  }
+  __finally {
+  }
+}
+
+// CHECK-LABEL: @"?seh_finally_with_obj@@YAXPEAUA@@@Z"(
+//
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     invoke void @llvm.seh.try.begin()
+//
+// CHECK:     invoke void @llvm.seh.scope.begin()
+// CHECK:     invoke void @"??1A@@QEAA@XZ"
+//
+// CHECK:     invoke void @llvm.seh.scope.end()
+//
+// CHECK-NOT: @llvm.seh.scope
+//
+// CHECK:     invoke void @llvm.seh.try.end()
+//
+// CHECK-NOT: @llvm.seh.scope
+// CHECK:     call void @"?fin$0@0@seh_finally_with_obj@@"(i8 noundef 0
+//
+// CHECK:     invoke void @llvm.seh.scope.end()
+// CHECK:     cleanupret
+//
+// CHECK:     call void @"?fin$0@0@seh_finally_with_obj@@"(i8 noundef 1
+// CHECK-NOT: @llvm.seh
+// CHECK:     cleanupret
+void seh_finally_with_obj(A* p) {
+  __try {
+    delete p;
+  }
+  __finally {
   }
 }

--- a/clang/test/CodeGenCXX/exceptions-seh.cpp
+++ b/clang/test/CodeGenCXX/exceptions-seh.cpp
@@ -16,6 +16,10 @@
 // RUN:         -fms-extensions -x c++ -emit-llvm-only -verify %s -DERR_SYNC2
 // RUN: %clang_cc1 -triple x86_64-windows -fcxx-exceptions -fexceptions \
 // RUN:         -fms-extensions -x c++ -emit-llvm-only -verify %s -DERR_SYNC3
+// RUN: %clang_cc1 -triple x86_64-windows -fasync-exceptions -fcxx-exceptions -fexceptions \
+// RUN:         -fms-extensions -x c++ -emit-llvm-only -verify %s -DERR_NEW_THROWING_CTOR
+// RUN: %clang_cc1 -triple x86_64-windows -fasync-exceptions -fcxx-exceptions -fexceptions \
+// RUN:         -fms-extensions -x c++ -emit-llvm-only -verify %s -DNOERR_NEW_NOEXCEPT_CTOR
 // RUN: %clang_cc1 -triple x86_64-windows \
 // RUN:         -fms-extensions -x c++ -emit-llvm-only -verify %s -DNOERR
 
@@ -208,6 +212,27 @@ void seh_unwinding() {
 void seh_unwinding() {
   HasCleanup x; // expected-error{{'__try' is not permitted in functions that require object unwinding}}
   __try {
+  } __except (1) {
+  }
+}
+#elif defined(ERR_NEW_THROWING_CTOR)
+void seh_unwinding() {
+  __try {
+    HasCleanup *p = new HasCleanup; // expected-error{{'__try' is not permitted in functions that require object unwinding}}
+    delete p;
+  } __except (1) {
+  }
+}
+#elif defined(NOERR_NEW_NOEXCEPT_CTOR)
+// new-expression with a noexcept constructor should be fine.
+struct NoThrowCtor {
+  NoThrowCtor() noexcept;
+  ~NoThrowCtor();
+};
+void seh_unwinding() {
+  __try {
+    NoThrowCtor *p = new NoThrowCtor; // expected-no-diagnostics
+    delete p;
   } __except (1) {
   }
 }

